curvefs/src/client/warmup/warmup_manager.cpp

```diff
@@ -0,0 +1,690 @@
+            inode2FetchS3ObjectsPoolMutex_.RDLock();
+        } else {
+            ++iter;
+        }
+    }
+    inode2FetchS3ObjectsPoolMutex_.Unlock();
+}
+
+void WarmupManagerS3Impl::ScanCleanWarmupProgress() {
+    // clean done warmupProgress
+    inode2ProgressMutex_.RDLock();
+    for (auto iter = inode2Progress_.begin(); iter != inode2Progress_.end();) {
+        if (ProgressDone(iter->first)) {
+            VLOG(9) << "warmup key: " << iter->first << " done!";
+            inode2ProgressMutex_.Unlock();
+            inode2ProgressMutex_.WRLock();
+            iter = inode2Progress_.erase(iter);
+            inode2ProgressMutex_.Unlock();
+            inode2ProgressMutex_.RDLock();
+        } else {
+            ++iter;
+        }
+    }
+    inode2ProgressMutex_.Unlock();
+}
+
+void WarmupManagerS3Impl::ScanWarmupFiles() {
+    // file need warmup
+    WriteLockGuard lock(warmupInodesDequeMutex_);
+    if (!warmupInodesDeque_.empty()) {
+        WarmupInodes inodes = warmupInodesDeque_.front();
+        for (auto const &iter : inodes.GetReadAheadFiles()) {
+            VLOG(9) << "BackGroundFetch: key: " << inodes.GetKey()
+                    << " inode:" << iter;
+            FetchDataEnqueue(inodes.GetKey(), iter);
+            warmupInodesDeque_.pop_front();
+        }
+    }
+}
+
+void WarmupManagerS3Impl::ScanWarmupFilelist() {
+    // Use a write lock to ensure that all parsing tasks are added.
+    WriteLockGuard lock(warmupFilelistDequeMutex_);
+    if (!warmupFilelistDeque_.empty()) {
+        WarmupFilelist warmupFilelist = warmupFilelistDeque_.front();
+        VLOG(9) << "warmup ino: " << warmupFilelist.GetKey()
+                << " len is: " << warmupFilelist.GetFileLen();
+
+        std::vector<std::string> warmuplist;
+        GetWarmupList(warmupFilelist, &warmuplist);
+        for (auto filePath : warmuplist) {
+            FetchDentryEnqueue(warmupFilelist.GetKey(), filePath);
+        }
+        warmupFilelistDeque_.pop_front();
+    }
+}
+
+void WarmupManagerS3Impl::AddFetchDentryTask(fuse_ino_t key,
+                                             std::function<void()> task) {
+    VLOG(9) << "add fetchDentry task: " << key;
+    std::unique_ptr<ThreadPool> tp = absl::make_unique<ThreadPool>();
+    tp->Start(warmupThreadsNum_);
+    inode2FetchDentryPoolMutex_.WRLock();
+    auto iter = inode2FetchDentryPool_.emplace(key, std::move(tp));
+    inode2FetchDentryPoolMutex_.Unlock();
+    if (!iter.first->second->Enqueue(task)) {
+        LOG(ERROR) << "key:" << key
+                   << " fetch dentry thread pool has been stoped!";
+    }
+
+    VLOG(9) << "add fetchDentry task: " << key << " finished";
+}
+
+void WarmupManagerS3Impl::AddFetchS3objectsTask(fuse_ino_t key,
+                                                std::function<void()> task) {
+    VLOG(9) << "add fetchDentry task: " << key;
+    std::unique_ptr<ThreadPool> tp = absl::make_unique<ThreadPool>();
+    tp->Start(warmupThreadsNum_);
```

Comment:
Each inode has a thread pool. What if there are many files downloaded?

Reply:
> Each inode has a thread pool. What if there are many files downloaded?

A warmup task has a thread pool, and if there are many files, share the thread pool

Reply:
> Each inode has a thread pool. What if there are many files downloaded?

A warmup task has a thread pool, and if there are many files, share the thread pool

Reply:
ok, need limit the task nums(thread nums)

Reply:
> > Each inode has a thread pool. What if there are many files downloaded?
> 
> A warmup task has a thread pool, and if there are many files, share the thread pool

threads will be created even same key

Reply:
> threads will be created even same key

that is todo work